### PR TITLE
Add hooks implementation

### LINKED ISSUE #148: Add hooks to API
Opening this issue for tracking purposes.

Things I've figured out so far:

* The notion of a two-pass render is kind of strange the context of hooks. I'm doing my best to approximate it, but it may come at the expense of two `setStates` in the same cycle when it's in frontend-only mode.
* Hooks don't play nicely with the current way that tests are written. I'm going to rewrite the tests to use the React Testing Library.

### PR DESCRIPTION
**Pull request by tstirrat15**
Fixes #148 

## Motivation
This has been requested functionality for a while now and I found the time to get around to it.

## Changes
I'll annotate the files, since this diff is pretty noisy.

## Notes
I'm not entirely sure that the two-pass SSR render behavior has been preserved. I'd love some help figuring that out in code review.

This should be released as a major version according to semantic versioning rules because peer dependencies are changing. It also ended up being a pretty major rewrite.

### CODE REVIEW

File: modules/__tests__/Media-test.js [hunk 1]
```diff
@@ -441,25 +392,21 @@ describe("A <Media> in browser environment", () => {
...
             {({ matches }) =>
               matches ? <div>fully matched</div> : <div>not matched</div>
             }
           </Media>
         );
 
-        renderStrict(element, node, () => {
-          expect(node.innerHTML).toMatch("not matched");
-        });
+        expect(container.innerHTML).toMatch("not matched");
```
Comment (tstirrat15) [inline modules/__tests__/Media-test.js:408]: Note that none of the test logic changed substantially.

On a side note, is it possible to test the two-pass render behavior? I don't have a solid enough understanding of how it should work to be able to write a test against it.

File: package.json [hunk 2]
```diff
@@ -39,6 +39,8 @@
     "@babel/plugin-transform-runtime": "^7.1.0",
     "@babel/preset-env": "^7.1.6",
     "@babel/preset-react": "^7.0.0",
+    "@testing-library/jest-dom": "^5.7.0",
+    "@testing-library/react": "^10.0.4",
```
Comment (tstirrat15) [inline package.json:47]: Adding React Testing Library and its buddies

File: package.json [hunk 3]
```diff
@@ -24,7 +24,7 @@
     "testTypes": "tsc --noEmit --jsx react typeTest.tsx"
   },
   "peerDependencies": {
-    "react": ">=15 || ^0.14.7"
+    "react": "^16.8.0"
```
Comment (tstirrat15) [inline package.json:27]: Adding a hook means that this is now stuck with 16.8. I made corresponding changes to the devdeps.

File: package.json [hunk 4]
```diff
@@ -39,33 +39,32 @@
     "@babel/plugin-transform-runtime": "^7.1.0",
     "@babel/preset-env": "^7.1.6",
     "@babel/preset-react": "^7.0.0",
+    "@rollup/plugin-babel": "5.0.0",
```
Comment (tstirrat15) [inline package.json:42]: `npm audit` complained about `rollup-plugin-snapshot-size` and upgraded it beyond a version that worked with the current version of `rollup`, so I upgraded Rollup and all its friends.

File: rollup.config.js [hunk 5]
```diff
@@ -1,9 +1,9 @@
...
-import replace from 'rollup-plugin-replace';
-import commonjs from 'rollup-plugin-commonjs';
-import nodeResolve from 'rollup-plugin-node-resolve';
+import babel from '@rollup/plugin-babel';
+import replace from '@rollup/plugin-replace';
+import commonjs from '@rollup/plugin-commonjs';
+import nodeResolve from '@rollup/plugin-node-resolve';
 import { sizeSnapshot } from 'rollup-plugin-size-snapshot';
-import { uglify } from 'rollup-plugin-uglify';
+import { terser } from 'rollup-plugin-terser';
```
Comment (tstirrat15) [inline rollup.config.js:6]: Uglify became Terser. These are otherwise functionally identical packages.

File: rollup.config.js [hunk 6]
```diff
@@ -18,35 +18,35 @@ function external(id) {
 const cjs = [
   {
     input,
-    output: { file: `cjs/${pkg.name}.js`, format: 'cjs' },
+    output: { file: `cjs/${pkg.name}.js`, format: 'cjs', exports: 'named' },
```
Comment (tstirrat15) [inline rollup.config.js:21]: Rollup wants you to be explicit about named exports when you're providing both a default and named export. It still complains if you provide both a default and a named export, but we're ignoring that for now.

File: rollup.config.js [hunk 7]
```diff
@@ -18,35 +18,35 @@ function external(id) {
...
     external,
     plugins: [
-      babel({ exclude: /node_modules/ }),
+      babel({ exclude: /node_modules/, babelHelpers: 'bundled' }),
```
Comment (tstirrat15) [inline rollup.config.js:24]: The new version of the babel plugin uses a different setting for whether to include the babel runtime, so this changed throughout the fiel.

File: .size-snapshot.json [hunk 8]
```diff
@@ -1,26 +1,26 @@
...
     "treeshaked": {
       "rollup": {
-        "code": 3016,
-        "import_statements": 322
+        "code": 364,
+        "import_statements": 187
       },
       "webpack": {
-        "code": 4345
+        "code": 3673
```
Comment (tstirrat15) [inline .size-snapshot.json:12]: Check it out! We're actually getting a smaller package out of this.

File: package.json [hunk 9]
```diff
@@ -24,48 +24,45 @@
     "testTypes": "tsc --noEmit --jsx react typeTest.tsx"
   },
   "peerDependencies": {
-    "react": ">=15 || ^0.14.7"
+    "react": "^16.8.0"
   },
   "dependencies": {
-    "@babel/runtime": "^7.2.0",
+    "@babel/runtime": "^7.9.6",
```
Comment (tstirrat15) [inline package.json:30]: Will this cause problems? Or will it just ensure that the runtime required by an end-user ends up being at least this recent?

**Comment by tstirrat15** [comment]
I was running into this error:
```
 FAIL  modules/__tests__/Media-test.js
  ● Test suite failed to run

    Requires Babel "^7.0.0-0", but was loaded with "6.26.3". If you are sure you have a compatible version of @babel/core, it is likely that something in your build process is loading the wrong version. Inspect the stack trace of this error to look for the first entry that doesn't mention "@babel/core" or "babel-core" to see what is calling Babel.

      at throwVersionError (node_modules/@babel/helper-plugin-utils/lib/index.js:65:11)
      at Object.assertVersion (node_modules/@babel/helper-plugin-utils/lib/index.js:13:11)
      at node_modules/@babel/plugin-proposal-export-default-from/lib/index.js:17:7
      at node_modules/@babel/helper-plugin-utils/lib/index.js:19:12
          at Array.map (<anonymous>)
```
And couldn't figure out what to make of it. I found a bunch of issues that talked about it:

* https://github.com/facebook/jest/issues/6992
* https://github.com/facebook/jest/issues/6662
* https://github.com/facebook/jest/issues/6913

but none of the fixes detailed in those issues seemed to work. I tried swapping `yarn` for `npm` on a whim, and it started working for reasons that I can't grok.

Any objections to using Yarn?

**Comment by tstirrat15** [comment]
This should be ready for review now.

**Comment by edorivai** [comment]
Good catch on trying yarn. I suppose there is some locked package that imports the wrong babel version for some reason. So I deleted the `package-lock.json` and re-ran `npm i`. After that `npm test` runs successfully, it does show some errors in the console though, I think this is expected. Perhaps you could confirm?

I actually think library authors shouldn't use `package-lock.json` since users won't be applying our lockfile when importing `react-media`.

```
> jest

 PASS  modules/__tests__/Media-ssr-test.js
 PASS  modules/__tests__/Media-test.js
  ● Console

    console.error node_modules/jsdom/lib/jsdom/virtual-console.js:29
      Error: Uncaught [Invariant Violation: <Media targetWindow> does not support `matchMedia`.]
          at reportException (/home/edo/work/react-media/node_modules/jsdom/lib/jsdom/living/helpers/runtime-script-errors.js:66:24)
          at invokeEventListeners (/home/edo/work/react-media/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:209:9)
          at HTMLUnknownElementImpl._dispatch (/home/edo/work/react-media/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:119:9)
          at HTMLUnknownElementImpl.dispatchEvent (/home/edo/work/react-media/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:82:17)

...more truncade stack traces
```

**Comment by edorivai** [comment]
Pinging @mjackson to give him a chance to review this if he has time. If he doesn't jump in we'll release an alpha soon.

**Comment by tstirrat15** [comment]
Ah, yeah. Looks like you're correct about lockfiles: https://dev.to/gajus/stop-using-package-lock-json-or-yarn-lock-3ddi

I'll remove the lockfile from this PR and see how it works.

I'll also do some research on silencing that error. I think it's in that test case that's designed to throw, and there's probably something I need to do in Jest to make sure that the error is fully caught.

**Comment by tstirrat15** [comment]
I've made both of those changes. I think there's some additional refactoring and simplification that could be done, but I'd be okay with leaving that for future work.

**Comment by edorivai** [comment]
Hey man, just letting you know I haven't forgotten about this! Life a bit hectic ATM, I'll do my best to squeeze this in ASAP.

**Comment by tstirrat15** [comment]
No worries! You've already been more present and responsive than multiple projects that I've submitted PRs to in recent months. Thank you!

**Comment by edorivai** [comment]
Now available at `react-media@next` :tada:

**Pull request merged**